File: Services/FeishuBridge/test/inbound-media.test.js
const assert = require('node:assert/strict');
const fs = require('node:fs');
const os = require('node:os');
const path = require('node:path');
const test = require('node:test');
const { assertPrivateMode } = require('./helpers/platform-private');
const {
  cleanupInboundAssets,
  inboundAssetsPrompt,
  inboundMessageDetails,
  stageInboundMessage,
} = require('../lib/inbound-media');

function temporaryDirectory(t) {
  const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'feishu-inbound-media-'));
  t.after(() => fs.rmSync(dir, { recursive: true, force: true }));
  return dir;
}

test('inbound message parsing extracts bounded resources without exposing keys in prompts', () => {
  const post = inboundMessageDetails({
    message_type: 'post',
    content: JSON.stringify({
      title: '周报',
      content: [[
        { tag: 'text', text: '请总结' },
        { tag: 'img', image_key: 'img_private' },
        { tag: 'img', image_key: 'img_private' },
      ]],
    }),
  });
  assert.equal(post.text, '周报\n请总结');
  assert.equal(post.resources.length, 1);
  assert.equal(post.resources[0].resourceType, 'image');
  const prompt = inboundAssetsPrompt(post.text, {
    assets: [{ messageType: 'post', resourceType: 'image', displayName: 'image', localPath: '/tmp/image', sizeBytes: 3 }],
  });
  assert.match(prompt, /\/tmp\/image/);
  assert.doesNotMatch(prompt, /img_private/);
});

test('inbound resources are downloaded to a private scoped directory and cleaned after use', async (t) => {
  const cwd = temporaryDirectory(t);
  const calls = [];
  const lark = {
    async larkImDownloadMessageResource(input) {
      calls.push(input);
      const output = path.join(cwd, input.output);
      fs.writeFileSync(output, 'private attachment');
      return {};
    },
  };
  const staged = await stageInboundMessage({
    lark,
    cwd,
    message: {
      message_id: 'om_private_message',
      message_type: 'file',
      content: JSON.stringify({ file_key: 'file_private_key', file_name: '../../report.txt' }),
    },
  });
  assert.equal(staged.assets.length, 1);
  assert.equal(staged.assets[0].displayName, 'report.txt');
  assert.equal(staged.assets[0].localPath.startsWith(path.join(cwd, 'node_modules', '.cache')), true);
  assertPrivateMode(staged.cleanupDir, 0o700);
  assertPrivateMode(staged.assets[0].localPath, 0o600);
  assert.equal(calls[0].fileKey, 'file_private_key');
  assert.equal(calls[0].output.includes('..'), false);
  cleanupInboundAssets(cwd, staged.cleanupDir);
  assert.equal(fs.existsSync(staged.cleanupDir), false);
});

test('inbound size limits fail closed and remove partial downloads', async (t) => {
  const cwd = temporaryDirectory(t);
  const lark = {
    async larkImDownloadMessageResource(input) {
      fs.writeFileSync(path.join(cwd, input.output), 'too large');
      return {};
    },
  };
  await assert.rejects(() => stageInboundMessage({
    lark,
    cwd,
    maxBytes: 4,
    message: {
      message_id: 'om_large',
      message_type: 'image',
      content: JSON.stringify({ image_key: 'img_large' }),
    },
  }), /size limit/);
  const root = path.join(cwd, 'node_modules', '.cache', 'feishu-bridge', 'inbound-assets');
  assert.deepEqual(fs.existsSync(root) ? fs.readdirSync(root) : [], []);
});

File: Core/internal/feishu/inbound_media_test.go
package feishu

import (
	"context"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"
)

func TestInboundMessageDetailsExtractsPostResourcesWithoutKeysInPrompt(t *testing.T) {
	message := InboundMessage{MessageID: "om_test", MessageType: "post", Raw: map[string]any{"message": map[string]any{"content": `{"title":"周报","content":[[{"tag":"text","text":"请总结"},{"tag":"img","image_key":"img_private"},{"tag":"img","image_key":"img_private"}]]}`}}}
	text, resources, err := inboundMessageDetails(message)
	if err != nil {
		t.Fatal(err)
	}
	if text != "周报\n请总结" || len(resources) != 1 || resources[0].ResourceType != "image" {
		t.Fatalf("unexpected details: %q %#v", text, resources)
	}

}

func TestStageInboundMessageUsesPrivateScopedDirectoryAndCleans(t *testing.T) {
	root := t.TempDir()
	message := InboundMessage{MessageID: "om_private", MessageType: "file", Raw: map[string]any{"message": map[string]any{"content": `{"file_key":"file_private","file_name":"../../report.txt"}`}}}
	staged, err := StageInboundMessage(context.Background(), root, fakeResourceDownloader{}, message, 1024)
	if err != nil {
		t.Fatal(err)
	}
	if len(staged.Assets) != 1 || staged.Assets[0].DisplayName != "report.txt" || !strings.HasPrefix(staged.Assets[0].LocalPath, filepath.Join(root, "private-cache", "inbound-assets")) {
		t.Fatalf("unexpected staged asset: %#v", staged)
	}
	info, err := os.Stat(staged.Assets[0].LocalPath)
	if err != nil {
		t.Fatal(err)
	}
	if runtime.GOOS != "windows" && info.Mode().Perm() != 0o600 {
		t.Fatalf("asset is not private: %#o", info.Mode().Perm())
	}
	if err := CleanupInboundAssets(root, staged.CleanupDir); err != nil {
		t.Fatal(err)
	}
	if _, err := os.Stat(staged.CleanupDir); !os.IsNotExist(err) {
		t.Fatal("staged directory was not removed")
	}
}

func TestStageInboundMessageRejectsOversizeAndRemovesPartialData(t *testing.T) {
	root := t.TempDir()
	message := InboundMessage{MessageID: "om_large", MessageType: "image", Raw: map[string]any{"message": map[string]any{"content": `{"image_key":"img_large"}`}}}
	if _, err := StageInboundMessage(context.Background(), root, fakeResourceDownloader{}, message, 4); err == nil {
		t.Fatal("oversize resource was accepted")
	}
	assetRoot := filepath.Join(root, "private-cache", "inbound-assets")
	entries, _ := os.ReadDir(assetRoot)
	if len(entries) != 0 {
		t.Fatalf("partial asset data remains: %#v", entries)
	}
}

type fakeResourceDownloader struct{}

func (fakeResourceDownloader) DownloadMessageResource(ctx context.Context, messageID, fileKey, kind, output string, timeout time.Duration) error {
	return os.WriteFile(output, []byte("private attachment"), 0o600)
}
